### PR TITLE
Feature #1246 - Remote syslog

### PR DESCRIPTION
**Pull request by wendigo**
This change introduces ability to log to remote syslog server.

I've made small refactoring to share code between ```errors``` and ```log``` plugins.

### CODE REVIEW

**Comment by mholt** [comment]
I'll look at this soon, very swamped with things!

**Comment by wendigo** [comment]
Fully understand :)

**Comment by mholt** [comment]
I'm still busy... sorry. Would somebody else like to review this if they have a chance before I do?

File: caddyhttp/httpserver/logger.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	"strings"
+)
+
+var remoteSyslogPrefixes = map[string]string{
+	"syslog+tcp://": "tcp",
+	"syslog+udp://": "udp",
+	"syslog://":     "udp",
+}
+
+type Logger struct {
```
Comment (mholt) [inline caddyhttp/httpserver/logger.go]: Check the golint output -- will need godoc on the exported things. ;)

File: caddyhttp/httpserver/logger.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+	Output string
+	*log.Logger
+	Roller *LogRoller
+	writer io.Writer
+}
+
+func NewTestLogger(buffer *bytes.Buffer) *Logger {
```
Comment (abiosoft) [inline caddyhttp/httpserver/logger.go]: This will also need godoc as long as it is not in a `_test.go` file.

**Comment by mholt** [comment]
Fabulous. Thank you @wendigo! Before I click Merge, can you provide some docs that I can put on the website? Either a PR on the caddyserver.com repo or just a comment here explaining precisely how it is used would be fine. I want to be sure I don't miss anything.

**Comment by mholt** [comment]
@wendigo I imagine you're pretty busy. When do you think you'll be able to document this? (Again, it doesn't have to go into the site directly, just write something up for us to get us started.)

**Comment by wendigo** [comment]
Yeah, I'll write some docs @mholt

**Comment by wendigo** [comment]
Hi @mholt!
Changes in usage are as follow:

both `log` and `errors` adds support for following log "locations":

```syslog``` - logs to local syslog
```syslog://host[:port]``` - logs via UDP to local or remote syslog server
```syslog+udp://host[:port]``` - logs via UDP to local or remote syslog server
```syslog+tcp://host[:port]``` - logs via TCP to local or remote syslog server

Final syntax is i.e.:

```log syslog://remotehost:442 {``` and
```errors syslog+udp://remotehost:442 {``` and
```
errors {
  log syslog://remotehost:442
}
```

Basically - everywhere where ```[logfile]``` can be passed proper syslog scheme can be :)

**Comment by wendigo** [comment]
I've also merged your change with RWMutex

**Comment by wendigo** [comment]
AppVeyor failed on `test -race TestStandaloneTLSTicketKeyRotation`

**Comment by tobya** [comment]
@wendigo Sorry to ask a silly question but did you merge the doc changes into the docs?  

If not I could look at creating the changes to docs for caddyserver.com based on your notes, unless someone beats me to it.

**Comment by mholt** [comment]
I was also just looking at this now that I'm catching up with life a bit :) If someone can prepare a PR for the docs on the website that would be splendid, and we'll merge this ASAP.

**Comment by mholt** [comment]
Thanks for your patience and effort @wendigo!

**Comment by tobya** [comment]
@wendigo @mholt syslog doesnt seem to be implemented on windows

https://golang.org/pkg/log/syslog/#pkg-note-BUG

I cant build.  If I got some suggestions around what to do (eg an if block, use another library etc) I could take a look, as is however though I think this won't build on windows (unless I'm missing something)

http://stackoverflow.com/questions/42083059/getting-syslog-writer-undefined

**Comment by wendigo** [comment]
Hi, I'm using https://github.com/hashicorp/go-syslog?files=1 which wrapps suslog using build tags

**Comment by wendigo** [comment]
Did you fetch it?

**Comment by tobya** [comment]
I just built it.  Should I fetch separately?

**Comment by tobya** [comment]
@wendigo Thank you I assumed go build would fetch latest but obviously not.  Builds fine now.

**Comment by wendigo** [comment]
@tobya good to hear that. Thx!

**Comment by mholt** [comment]
Phew, you guys scared me for a while. 😅

**Pull request merged**